sdk/core/src/error/mod.rs

```diff
@@ -0,0 +1,354 @@
+use std::borrow::Cow;
+use std::fmt::{Debug, Display};
+
+mod azure_core_errors;
+mod hyperium_http;
+mod macros;
+
+/// A convience alias for `Result` where the error type is hard coded to `Error`
+pub type Result<T, O = std::convert::Infallible> = std::result::Result<T, Error<O>>;
+
+/// The kind of error
+///
+/// The classification of error is intentionally fairly coarse.
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum ErrorKind<O> {
+    /// An error specific to a certain operation
+    Operation(O),
+    /// An HTTP status code that was not expected
+    UnexpectedOperation { status: u16 },
```

Comment:
Suggested name `ResponseFailed` from @JeffreyRichter.

The user might also need the full HttpResponse for debugging purposes. Additionally there is a "x-ms-error-code" that can be read from a header (or the body if the header happens to be missing). This should be made available to the user of this library. 

More info here: https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#handling-errors

Reply:
I prefer `RequestFailed`. The caller's request failed for some reason. The response contains that reason, just like a failed function returns an error code. But to compare, here's what we use in other languages:

* .NET: `RequestFailedException`
* Java: `HttpResponseException`. There is a `HttpRequestException` but does seem to be geared toward a failed request. They also have quite a few exceptions for different responses, similar to what this rust SDK does now but you're changing.
* JavaScript: `RestError`
* Python: `HttpResponseError`
* C++: `RequestFailedException`

So it does seem like a pretty inconsistent mix of "request" and "response".